cps/oauth_bb.py

```diff
@@ -22,328 +21,183 @@
+
+        # log.debug("authorized %s: %s token %s", provider.id, current_user, token)
         if not token:
-            flash(_(u"Failed to log in with Google."), category="error")
+            flash(_(u"Failed to log in with %s." % provider.name), category="error")
             return False
 
-        resp = blueprint.session.get("/oauth2/v2/userinfo")
-        if not resp.ok:
-            flash(_(u"Failed to fetch user info from Google."), category="error")
+        account_id = provider.get_account_id()
+        if not account_id:
+            flash(_(u"Failed to fetch user info from %s." % provider.name), category="error")
             return False
 
-        google_info = resp.json()
-        google_user_id = str(google_info["id"])
-        return oauth_update_token(str(oauthblueprints[1]['id']), token, google_user_id)
-
-
-    def oauth_update_token(provider_id, token, provider_user_id):
-        session[provider_id + "_oauth_user_id"] = provider_user_id
-        session[provider_id + "_oauth_token"] = token
-
-        # Find this OAuth token in the database, or create it
-        query = ub.session.query(ub.OAuth).filter_by(
-            provider=provider_id,
-            provider_user_id=provider_user_id,
-        )
-        try:
-            oauth = query.one()
-            # update token
-            oauth.token = token
-        except NoResultFound:
-            oauth = ub.OAuth(
-                provider=provider_id,
-                provider_user_id=provider_user_id,
-                token=token,
-            )
-        try:
-            ub.session.add(oauth)
-            ub.session.commit()
-        except Exception as e:
-            log.exception(e)
-            ub.session.rollback()
+        result = provider.confirm_token(account_id, token)
+        # log.debug("authorized %s: %s updated token => %s", provider.id, current_user, result)
+        if not result:
+            log.warning("auhorized %s: %s failed to update token", provider.id, current_user)
+            return False
 
-        # Disable Flask-Dance's default behavior for saving the OAuth token
-        return False
-
-
-    def bind_oauth_or_register(provider_id, provider_user_id, redirect_url):
-        query = ub.session.query(ub.OAuth).filter_by(
-            provider=provider_id,
-            provider_user_id=provider_user_id,
-        )
-        try:
-            oauth = query.one()
-            # already bind with user, just login
-            if oauth.user:
-                login_user(oauth.user)
-                return redirect(url_for('web.index'))
+        provider.set_session_error(False)  # clear eventual error stored in the session
+        if hasattr(result, 'nickname'):  # type: ub.User
+            if current_user.is_authenticated:
+                # log.debug("authorized %s: current_user %s", blueprint.name, current_user)
+                assert current_user.id == result.id
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846306)